lib/vrl/compiler/src/expression.rs

```diff
@@ -69,6 +69,41 @@ pub use query::{Query, Target};
 pub use unary::Unary;
 pub use variable::Variable;
 
+/// A wrapper around an expression, that always resolves the `resolve_void`
+/// method.
+///
+/// This wrapper is used by the compiler to ensure an expression for which its
+/// return value can be omitted always does so.
+#[derive(Debug, Clone)]
+pub(crate) struct VoidExpr(pub(crate) Expr);
```

Comment:
> I'm not sure I fully understand the need for `VoidExpr`. I thought this PR was purely for the perf benefit, but this struct seems to be enforcing "correctness".

The VRL runtime _always_ calls `Expression::resolve`. The compiler merely makes sure that for any expression for which we can remove the return value, `resolve_void` is called instead, that is what this wrapper struct is about. It prevents the runtime from needing more information (and boolean switching) to know which variant to call for each resolved expression.

> Don't we just throw away the cloned `Value` if it's not used? Seems like this just swaps the value that would be thrown away with `null` instead?

Correct. This is what's explained in the PR description (and on `Expression::resolve_void`. For any non-optimized expression, this has zero impact on performance, but for the ones that do implement a custom `resolve_void`, they already avoid the extra clone.

Now, because I don't (yet) want to have a massive rewrite of expressions and how the regular `resolve` works (i.e. it is always expected to return a `Value`), this solution then returns `Value::Null`, which stores no internal data, so should be cheap to return.

Reply:
Got it. I missed the fact that the runtime doesn't choose which one to call. I assumed this was in addition to the compiler.
This part makes sense.